File: src/main/java/io/github/alxiw/icq/geekbot/api/ServiceProvider.java
package io.github.alxiw.icq.geekbot.api;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class ServiceProvider {

    private static final String BASE_URL = "https://geek-jokes.sameerkumar.website/";
    private static Service service;

    private ServiceProvider() {}

    public static Service getService(){
        if (service == null) {
            service = provideService();
        }
        return service;
    }

    private static Service provideService() {
        HttpLoggingInterceptor interceptor = provideInterceptor();
        OkHttpClient client = provideClient(interceptor);
        return new Retrofit.Builder()
                .baseUrl(BASE_URL)
                .client(client)
                .addConverterFactory(GsonConverterFactory.create())
                .build()
                .create(Service.class);
    }

    private static OkHttpClient provideClient(HttpLoggingInterceptor interceptor) {
        return new OkHttpClient.Builder()
                .addInterceptor(interceptor)
                .build();
    }

    private static HttpLoggingInterceptor provideInterceptor() {
        return new HttpLoggingInterceptor(HttpLoggingInterceptor.Logger.DEFAULT)
                .setLevel(HttpLoggingInterceptor.Level.BASIC);
    }
}

File: src/main/java/io/github/alxiw/hello/service/google/ServiceProvider.kt
package io.github.alxiw.hello.service.google

import io.github.alxiw.hello.sys.AppLogger
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import okhttp3.logging.HttpLoggingInterceptor.Logger
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

private const val BASE_URL = "https://translate.google.com/"

object ServiceProvider {

    private var service: GoogleService? = null

    fun getService(): GoogleService {
        return service ?: provideService().also { service = it }
    }

    private fun provideService(): GoogleService {
        val interceptor = provideHttpLoggingInterceptor()
        val client = provideClient(interceptor)
        return Retrofit.Builder()
            .baseUrl(BASE_URL)
            .client(client)
            .addConverterFactory(GsonConverterFactory.create())
            .build()
            .create(GoogleService::class.java)
    }

    private fun provideClient(interceptor: HttpLoggingInterceptor): OkHttpClient {
        return OkHttpClient.Builder()
            .addInterceptor(interceptor)
            .build()
    }

    private fun provideHttpLoggingInterceptor(): HttpLoggingInterceptor {
        return HttpLoggingInterceptor(object : Logger {
            override fun log(message: String) {
                AppLogger.i(message)
            }
        }).apply {
            level = HttpLoggingInterceptor.Level.BASIC
        }
    }
}
